joint_trajectory_controller/include/joint_trajectory_controller/stop_trajectory_builder.h

```diff
@@ -0,0 +1,145 @@
+//     this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//////////////////////////////////////////////////////////////////////////////
+#ifndef STOP_TRAJECTORY_BUILDER_H
+#define STOP_TRAJECTORY_BUILDER_H
+
+#include <vector>
+
+#include <joint_trajectory_controller/trajectory_builder.h>
+
+namespace joint_trajectory_controller
+{
+
+/**
+ * @brief Builder creating a trajectory stopping the robot.
+ *
+ * @note The start state of the stop motion and the final settle/hold state
+ * are eqal.
+ */
+template<class SegmentImpl>
+class StopTrajectoryBuilder : public TrajectoryBuilder<SegmentImpl>
+{
+private:
+  using Segment               = JointTrajectorySegment<SegmentImpl>;
+  using TrajectoryPerJoint    = std::vector<Segment>;
+  using Trajectory            = std::vector<TrajectoryPerJoint>;
+
+  using RealtimeGoalHandle    = realtime_tools::RealtimeServerGoalHandle<control_msgs::FollowJointTrajectoryAction>;
+  using RealtimeGoalHandlePtr = boost::shared_ptr<RealtimeGoalHandle>;
+
+public:
+  StopTrajectoryBuilder(const unsigned int& number_of_joints,
+                        const typename Segment::Time& stop_traj_duration,
+                        const typename Segment::State& hold_state);
+
+public:
+  /**
+   * @brief Creates a trajectory which reaches the settle position in a fixed time.
+   *
+   * The calculation is done as follows:
+   * - Create segment that goes from current (pos,vel) to (pos,-vel) in 2x the desired stop time.
+
+   * - Assuming segment symmetry, sample segment at its midpoint (desired stop time). It should have zero velocity.
+   * - Create segment that goes from current state to above zero velocity state, in the desired time.
+   *
+   * @note The symmetry assumption from the second point above might not hold for all possible segment types.
+   *
+   * @returns true if the building of the stop trajectory succeeds, otherwise false.
+   */
+  bool buildTrajectory(Trajectory* hold_traj) override;
+
+private:
+  const unsigned int number_of_joints_;
+  const typename Segment::Time stop_traj_duration_;
+  // To always have the newest value of the hold state,
+  // a reference is stored.
```

Comment:
Change the comment into a doxygen documentation and maybe write something like:
"To always have the newest value of the hold state, a reference is stored to the state which has to be reached when the stop-trajectory has to be build via buildTrajecory() function."

Reply:
Done.